### ISSUE TITLE
inconsistencies in truncation vs rounding

### ISSUE THREAD

**Issue by edgarcosta** [LMFDB/lmfdb#2632]
When displaying floats, should we round or truncate?
Neither is correct, but we should stick to one.

At the moment we do both, and I think we should round.
For example, in Maass forms and L-functions, we truncate almost all the time, with the exception if number rounds to an half-integer, then we then we take that approximation.
To me this doesn't make much sense.
We are happy to display 1 when given 0.99999... and 0.5 when given 0.49999..., but we aren't happy to display 0.1 when given 0.09999...?

ps: this isssue has been brought up before here: #1978, when we dediced to round 0.49999 to 0.5 and here #1293 regarding the zera zeros

**Comment by davidfarmer** [comment]
I think we should truncate.  The text we display is meant for the
eyes of a human.  Humans (me, anyway) want to see correct information.
\pi = 3.1415... .

Note the ... .  The ... and the = are the key.  It is wrong to write
"=" and not have the ... (although maybe the ... is implied).

I am pretty sure that nobody can disagree with this:
    If we write ... after the last digit we show, then we need
    to truncate because the ... stands for digits which have been hidden.

If we are going to round and not use ..., then we need another symbol
than "=", or else the statement is just plain wrong.

I prefer to see

\pi = 3.1415... .

than

\pi \approx 3.1416 .

Does anyone have a 3rd proposal for how to show \pi to 4 decimal
places?  Faced with those two options, I reject the second one because
\approx is not well defined.  We can try to define it to mean "all
digits shown are correct except the last one may be rounded", but I
don't see any real benefit in doing that. Unless I am missing something,
the first is correct and unambiguous.

Two more things:

1) When producing a number for use by a machine, there are good arguments
for rounding instead of truncating.  (But I don't think that is the use
case under discussion.)  And also good arguments for including possibly
incorrect digits after the asserted precision.

2) If the true value is 0.0999999..., then that really equals 0.1, and
it is not a question of rounding.  If the true value is 0.9999996441...,
and we are showing 3 decimal places, then the arguments above say
that we should show 0.999... .

It is true that there are places where we know (or conjecture) that the
actual answer has to be in the set {..., -1/2, 0, 1/2, 1, ...}.
In that case, converting 0.4999.... to 0.5 is not an act of rounding.

**Comment by edgarcosta** [comment]
> If we are going to round and not use ..., then we need another symbol
> than "=", or else the statement is just plain wrong.

I disagree.
If `x \in RR` and we write `x = 3.1416`, I  understand `round(x * 10^4) = 31416`.
If we wanted to say that `x = 31416/10^4`, then that is what we would have written.

If we go for the truncation display, then `x = 3.1416` means `floor(x * 10^4) = 31416`, this sounds less natural to me, given that we are doing arithmetic with 'round to nearest, ties to even'.
Furthermore, by doing arithmetic with 'round to nearest, ties to even' that is why we saw 0.499 in #1978 

> It is true that there are places where we know (or conjecture) that the
> actual answer has to be in the set {..., -1/2, 0, 1/2, 1, ...}.
> In that case, converting 0.4999.... to 0.5 is not an act of rounding.

The code doesn't know that, it simply applies the same rule to everything. 
As we have it, given 0.99999000... it will display 1.

**Comment by jwj61** [comment]
I think David's comments are sensible.  In particular, the use of equals and ... seem clear to me.

**Comment by davidfarmer** [comment]
I think you are proposing the 3rd option I asked about:

\pi = 3.1416

That requires us explaining that in some settings an LMFDB equals sign
has a different meaning that the usual equals sign.

Mis-applying existing code may be a problem, in which case there
should be an issue.  That code (which probably I wrote or contributed
to) is definitely being used in places where the intention is to
recognize that the 0.499... in the database is representing 1/2.

Another issue is what to do when we write 0.5, which appears to
be an approximation, when we actually mean 1/2 exactly.  I recall
some situation where a series had a mixture of exact and decimal
numbers, and I chose to use decimals everywhere.  (But I am straying
from the original point of this thread.)

> given that we are doing arithmetic with 'round to nearest, ties to even'.

I don't think how we do the arithmetic is relevant.  We are talking about
how to display numbers, where we are intentionally showing less precision
than actually available.

**Comment by edgarcosta** [comment]
Got it. 

I wrote some code to centralize the displaying of floating point numbers, and for now the default is to truncate.

Two questions:
* for special values of L-functions, do we round or do we truncate?

* do we want a special behavior for other rational numbers or just half integers?

regarding, your case in mind, I don't think we should have a mixture of exact and decimal numbers in the same expression.

**Comment by davidfarmer** [comment]
I think the special values should be treated the same as anything else.  But if you wanted to
write them as L(1/2) \approx rounded_decimal_with_no_dots, just as an experiment to see
how people react, that could be worth trying.  (But in a table or other places where there
is no explicit \approx symbol, then it either is truncation with dots, or else a supplementary
explanation of what is being shown.) 

I was only thinking of integers and half-integers for special treatment, but I am not familiar
with every place where decimals appear.

**Comment by JohnCremona** [comment]
On Thu, 27 Sep 2018 at 15:07, David W. Farmer <notifications@github.com>
wrote:

> I think the special values should be treated the same as anything else.
> But if you wanted to
> write them as L(1/2) \approx rounded_decimal_with_no_dots, just as an
> experiment to see
> how people react, that could be worth trying. (But in a table or other
> places where there
> is no explicit \approx symbol, then it either is truncation with dots, or
> else a supplementary
> explanation of what is being shown.)
>
> I was only thinking of integers and half-integers for special treatment,
> but I am not familiar
> with every place where decimals appear.
>

On elliptic curve pages there are several places where I use &approx;  and
a truncated/rounded decimal, for example on
http://beta.lmfdb.org/EllipticCurve/Q/38850/d/2 this is the case for the
canonical height of a point, a period and a regulator, and the special
value.

[There is also a bug I just noticed which is not in the old code at www, in
the display of Kodaira symbols where there are some extra backslashes.
Perhaps I should try to fix that since I have still not yet touched any
code since the postgres switch...]

**Comment by davidfarmer** [comment]
I think all I have been saying is that I prefer

\pi = 3.1415....

to

\pi \approx 3.1416

and that those are the only two reasonable options.

I am not against using each in different places.  But we need to be
aware that the second option has no universally accepted mathematical
meaning.  You may mean "the last decimal place is rounded".
And it would be good if we are consistent about that.  But readers may
think it means something else (or more likely, may not even think about
what it means).

**Comment by edgarcosta** [comment]
I think the intrinsic problem with truncation is that we are assuming that someone gave you the number truncated and did not round at some point, e.g:
```
sage: R = RealIntervalField(100)
....: foo = R(1/8 - 5*10**-19)
....: print "%s" % foo.str()
....: print "%f" % float(foo)
....: print "%f" % float(str(foo)[:-1])
....:
0.1249999999999999995000000000000?
0.125000
0.125000
```

In this case, even though foo was computed with 100 bits of precision, when we store in the database as a float, it gets rounded, even if we gave it the truncated string without the `?`.

**Comment by edgarcosta** [comment]
@davidfarmer what do you think about http://cmfs.lmfdb.xyz/L/Character/Dirichlet/19/8/ ? vs http://www.lmfdb.org/L/Character/Dirichlet/19/8/?

the major difference is to display always the same number of significant zeros, even if they are zeros on the right.

**Comment by davidfarmer** [comment]
I think that is a hard case.  This is exactly the situation where the
values are exactly 1/2, so I don't think the extra zeros make sense.

But if the number being truncated did not have any special properties,
then I think showing the trailing zeros is good and conveys useful
information.

**Comment by davidfarmer** [comment]
The coefficient of 5^{-s} on this page has trailing zeros, which I think looks good:

http://beta.lmfdb.org/L/ModularForm/GL3/Q/Maass/1/1/16.40312_0.171121/-0.4216864/

**Comment by AndrewVSutherland** [comment]
I think the Gamma_R(s+1.00) displayed in the functional equation on http://cmfs.lmfdb.xyz/L/Character/Dirichlet/19/8/ should be displayed as Gamma_R(s+1).

**Comment by edgarcosta** [comment]
Okay, I think now I understand the problem a bit better, and I will try again next week.

But the overall solution is to do the rounding at the time of loading the data (for some specific cases), and not at the time of printing, as that code should generic and doesn't know the precision of what was given.

**Comment by davidfarmer** [comment]
Yes, definitely Gamma_R(s+1)

The real part of the Gamma_R shift is always 0 or 1,
and the real part of a Gamma_C shift is always an integer
or half-integer.

**Comment by edgarcosta** [comment]
I will enforce that. Tks

**Comment by AndrewVSutherland** [comment]
I suggest that we make time to discuss this at the IAS workshop when we will all be in the same room.  We do need to agree on a set of standard conventions  for displaying real numbers in the LMFDB, because this issue is going to keep coming up.

**Comment by davidfarmer** [comment]
Discussing this in person is a good idea.  Here are the points I will 
make:

1.  When the number is going to be used by a machine, it is fine for
implementation issues to be given priority.  I think that means
rounding, and I am happy for those who understand the issues to make
that decision.

2.  If you expect a human to read the number, most people will find
it natural to assume the displayed digits are correct.

3.  The \approx symbol and trailing \ldots are the reasonable options
to indicate that a given decimal is an approximation.  The \ldots
makes it more obvious that the decimal was truncated.  I think \approx
is compatible with truncation.

4.  The thing I really care about is having home pages meet the
expectation of the majority of users.  Most of those are nonexperts,
which is why I argue for truncation.  Us experts can handle ether
rounding or truncations, as long as we know which is being shown.

Rounding and truncation specify a number to the same precision,
i.e., the interval being specific has the same width.

**Comment by JohnCremona** [comment]
I agree with all the above regarding what we display on web pages.  There is also an  issue about how we store floating point values in the database.  This has changed in the transition to postrgesql (as far as I can see), and there are various options possible and in use, from fixed strings to floats to a fancier RealLiteral type.  (I have probably misunderstood this.)  This is definitely something I want to understand by the end if the IAS workshop.  When I do, I will be reworking all the f.p. values stored with elliptic curves since (as revealed in the recent data quality review) some of these, despite being computed to some reasonably high precision, went through a variety of transformations (by me in various scripts and not all intentional) including bottlenecks such as python floats.  (This includes: heights of generators, regulator, periods, special values.)

**Comment by AndrewVSutherland** [comment]
@JohnCremona Sounds good.

**Comment by edgarcosta** [comment]
After our discussion, I propose that we *always* round the floating points whenever display them.
This would avoid issues like https://github.com/LMFDB/lmfdb/issues/1978 , which was fixed with a hack, where we decided to round to half-integers (and hence integers), but nothing else.

For the future we should follow @arbooker suggestion:

 

> I would like to move towards a world in which all real numbers in the LMFDB are stored as intervals.
> 
> My proposal is to store a pair (m,r), where m is a numeric encoding the midpoint (as a multi-precision decimal), and r is an 8-byte (binary) double encoding the radius.
> 
> For display purposes, the default behavior would be to display only the radius, but there should be an option for the user to see the interval (by mouseover/knowl/whatever). This would be written in the form m ± r', where r' is a "%g" decimal encoding of r, but rounded up in the last place.
> 
> Since we will store and display intervals as midpoint and radius, it would make sense to round m when converting to decimal; to do otherwise would involve artificially increasing the radius. Whichever convention we choose, the conversion to decimal would by done by the upload script (taking care to ensure that the radius is increased appropriately). The number of digits to store/display in m would also be decided at upload time; we may want fixed precision in some cases, and a variable precision chosen based on the size of the radius in others.
>

**Comment by jvoight** [comment]
Thanks!  I like both of these proposals.

**Comment by davidfarmer** [comment]
Suppose that X is the "complete" decimal representation of a number that will be shown
on an LMFDB page.  For the purpose of this discussion X already exists.  
How it was computed and how X was determined are irrelevant.  All that matters is that X contains
everything we know about that number (meaning all the possible digits, which is what I meant
by "complete".)  For example, X might be a zero of the Riemann zeta function to 30 digits,
where the last digit might be wrong.

Proposal (which I think is just my rephrasing of Andy's proposal):

If X is to be shown on a web page then just show X.  That is, all those digits.

If you choose to show fewer digits than all of X, then truncate.  For example, if the first
zero of zeta were shown to 6 decimal places, then we see
14.134725
If it were shown to 3 decimal places, then we see
14.134

The following thought experiment is persuasive for me:

Imagine there is a "show more digits" button and a "show fewer digits" button.
Repeatedly clicking either button in any order, just shows and hides digits:
it does not change any of the digits you see.

And maybe we will eventually have those buttons.

**Comment by roed314** [comment]
The issue is that "last digit might be wrong" doesn't just affect the last digit.  Suppose that we know a value to be `0.4999999999999999999999999999999999999999999` where the last digit might be wrong.  What that means is that the true value might be `0.5000000000000000000000000000000000000000000`.

**Comment by davidfarmer** [comment]
I think you are saying `X=0.4999999999999999999999999999999999999999999`
in that example, meaning that string of digits is the best decimal representation you have
for the quantity represented by X.

I take that to mean that if you were going to display all the information you have
about that quantity, then you would display X.  There are two parts to my proposal,
the first of which is that you just display all the digits you have, when you are trying
to give all the available information.  The documentation would make it clear that
the actual quantity is in an open interval that contains 1/2.

I'll proceed, assuming that we agree with the first part of my proposal.  (Otherwise,
skip the rest of what I am writing and describe what you don't like about that first part.)

The second part of my proposal is that if the situation on the web page only called
for 10 digits after the decimal, then it would be displayed as
`0.4999999999`.
In other words, exactly what would appear if you clicked the "show fewer digits" button.
I would not like to see `0.5` or `0.5000000000` in that case, any more than I would want to
see one of those if X equaled `0.4999999999998675309111111999999999999999`.

Note that my proposal is agnostic about how, prior to sending information to the
web page, you decide whether X equals the string
`0.4999999999999999999999999999999999999999999` or the string
`0.5000000000000000000000000000000000000000000`.  
(I call X a string, because when it is displayed it is just characters on a web page.)

My proposal is also agnostic about any implied accuracy.  Maybe the last 10 digits of X might
be wrong.  Fine with me, and my proposal does not handle that case differently.

One more thing about my thought experiment:  A "show less precision" button might
function differently than a "show fewer digits" button.  But I don't think it is as natural
for a person to decide to make things less precise.  A person it trying to absorb the
information on a page and is feeling overwhelmed.  So they want to hide some of
the information, not make the information worse.

**Comment by arbooker** [comment]
I agree with @davidfarmer on this. So in line with the proposal that I made, we decide the decimal representation of each number in advance, and that will normally be done by rounding. However, it might happen that we have (say) 30 decimal places of precision on some number, but we only want to display 6. In that case it would make sense to truncate. That way when the user hovers over the number to see the full interval version, the numbers will match up to 6 places.

**Comment by AndrewVSutherland** [comment]
I disagree with @davidfarmer and @arbooker.  It is not uncommon to see the exact same numerical figure printed to different precision in different publications (e.g. unemployment rates, median wages, census counts), and when this happens, journalists and their editors round, they do not truncate.  The public is quite used to this-- one commonly sees the disclaimer that the figures in a table may not sum to listed totals "due to rounding", but one never sees "due to truncating".   Rounding communicates more honestly, which is why both journalists and scientific publications do it (as Kiran commented at lunch, in the real world truncation is more often used to conceal information than to convey it, e.g. as with pricing).

So it would not bother me at all to see the numbers change when the display precision change, in fact I would expect this to happen.

One final comment and then I will stop (as I think we all should -- github is really not the place for this kind of debate).   We are often dealing with real numbers that have been computed in different ways, by different software systems, with different representation (binary/decimal, fixed precision/floating point).  It is not uncommon that the exact same real number will have been computed in two completely different ways (and associated to different mathematical objects).  Is entirely possible that a real number that happens to be exactly equal to 1/64 is stored as 0.126499999... to 64 bits of precision in one computation and 0.12650... to 100 decimal digits of precision in another.  When these two values are listed in successive rows in a table that is giving 6 digits of decimal precision, they should look the same.
This example is representative of the actual problem we are trying to solve (as opposed to the strawman question of how to display pi, where we already have an agreed solution: $\pi$).

**Comment by bmatschke** [comment]
Just to add how NumberDB is doing it (see [here](https://www.numberdb.org/help#section-number-types)): It distinguishes between two data types: rationals and reals. Rational numbers are presented exactly as fractions. Real numbers are presented in decimal expansion up to 1 ulp (unit of least precision). That is, 0.5 denotes the real interval [0.4,0.6], whereas 1/2 denotes an exact value.
I think that neither rounding nor truncation is always satisfactory: 

- We cannot promise that every displayed number is rounded: E.g. a numerically computed number 0.5000... cannot be safely rounded to 0 or 1. 
- Similarly, we cannot promise that every displayed number is truncated: E.g. a numerically computed number 1.00000... cannot be safely truncated to 0 or 1.

So my suggestion is: If a number is provably rational, then display a fraction. Otherwise, display it decimally up to 1 ulp (obtained either by rounding or truncation; say by rounding if it is perceived as more natural).
